### PR TITLE
Gradient Accumulation 2x: double effective batch size

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Larger effective batch sizes produce smoother gradient estimates, leading to better generalization. Our current batch size is 1 (each training step processes one sample — the meshes are large). By accumulating gradients over 2 samples before each optimizer step, we double the effective batch size to 2.

**Why this should help OOD:** Smoother gradients reduce the noise in the optimization trajectory, leading to flatter minima. Flatter minima generalize better to OOD inputs (Sharp Minima Generalization Theory — Keskar et al., 2017).

**Tradeoff:** With accumulation=2, we see half as many optimizer steps per epoch. But each step has lower gradient variance. With our 180-min timeout, the number of epochs should be nearly unchanged (the forward/backward passes dominate, not the optimizer step). We just do half as many optimizer.step() calls.

**Evidence:** Gradient accumulation is standard practice for transformers when memory limits batch size. For our case, we have memory to spare (46/96 GB used) but the mesh-per-sample is so large that batch_size=1 is natural. Accumulation=2 gives us the benefits of batch_size=2 without memory increase.

### Step 1: Add argument

```python
parser.add_argument('--grad_accum_steps', type=int, default=1,
                    help='Gradient accumulation steps (effective batch = batch_size * accum)')
```

### Step 2: Implement gradient accumulation

Modify the training loop:

```python
optimizer.zero_grad()
for step_i, batch in enumerate(train_loader):
    # Forward + backward
    loss = compute_loss(model, batch)
    loss = loss / args.grad_accum_steps  # Scale loss by accumulation steps
    loss.backward()
    
    # Only step optimizer every grad_accum_steps
    if (step_i + 1) % args.grad_accum_steps == 0:
        optimizer.step()
        optimizer.zero_grad()
        # Update EMA
        if ema_model is not None:
            update_ema(ema_model, model, args.ema_decay)
```

**Critical:** Scale the loss by 1/grad_accum_steps before backward(). This ensures the accumulated gradient has the correct magnitude regardless of accumulation factor.

**Critical:** EMA update should happen at the same frequency as optimizer.step() (every accum steps), not every forward pass.

### Step 3: Run 2 seeds with accum=2

```bash
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/accum2-s42" --seed 42 \
  --wandb_group gradient-accumulation \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling \
  --grad_accum_steps 2

# Repeat for seed 73: --seed 73 --wandb_name "alphonse/accum2-s73"
```

## Baseline

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in | 11.742 | < 11.742 |
| p_oodc | 7.643 | < 7.643 |
| p_tan | 27.874 | < 27.874 |
| p_re | 6.419 | < 6.419 |

- **Baseline W&B runs:** k5qwvce4 (seed 42), 7oa5xfhi (seed 73)
- **val/loss baseline:** ~0.37
- **Baseline effective batch size:** 1

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2335 because Gradient accumulation 2x significantly regresses all metrics: p_in +10.0%, p_oodc +12.1%, p_tan +7.6%, p_re +16.7%. Doubling effective batch size via 2-step gradient accumulation hurts generalization — the current batch size is already well-tuned and the model benefits from more frequent gradient updates rather than larger effective batches. W&B: xfx7x5d8 (s42), vgasoa4b (s73)..

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Seed 42 (W&B: `xfx7x5d8`)

| Metric | Baseline | Accum=2 | Delta |
|--------|----------|---------|-------|
| p_in | 11.742 | 12.201 | +3.9% |
| p_oodc | 7.643 | 8.643 | +13.1% |
| p_re | 6.419 | 7.597 | +18.4% |
| p_tan | 27.874 | 29.284 | +5.1% |
| val_loss | ~0.37 | 0.401 | +8.4% |

### Seed 73 (W&B: `vgasoa4b`)

| Metric | Baseline | Accum=2 | Delta |
|--------|----------|---------|-------|
| p_in | 11.742 | 13.547 | +15.4% |
| p_oodc | 7.643 | 8.270 | +8.2% |
| p_re | 6.419 | 7.532 | +17.3% |
| p_tan | 27.874 | 29.705 | +6.6% |
| val_loss | ~0.37 | 0.401 | +8.4% |

### Commands used
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent alphonse --wandb_name "alphonse/accum2-s42" --seed 42 \
  --wandb_group gradient-accumulation \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling \
  --grad_accum_steps 2
# Seed 73: --seed 73 --wandb_name "alphonse/accum2-s73" on GPU 1
```

- **Total epochs:** 149 (both seeds, hit 180-min timeout)
- **Epoch time:** ~79-80s (similar to baseline — accumulation adds negligible overhead)
- **Peak GPU memory:** ~84 GB (seed 42), ~65 GB (seed 73) — higher than baseline's ~46 GB, likely due to EMA model copies

### What happened

**The hypothesis did not work.** Gradient accumulation with 2x effective batch size degraded all metrics across both seeds:

- **Surface pressure (key metric):** Worse by 4-18% depending on split and seed. OOD-Re was hit hardest (+17-18%), suggesting the smoother gradients actually hurt the model's ability to generalize to unseen Reynolds numbers.
- **val_loss:** Consistently ~0.40 vs baseline ~0.37, a clear regression.
- **Both seeds agree:** The degradation is consistent across seeds, not a random fluctuation.

**Why it likely failed:** With batch_size=1, each sample is a large mesh (~55k nodes). The gradient from a single sample is already quite informative — it's not noisy in the way that small-image-batch gradients are in vision tasks. Accumulating over 2 samples effectively smooths out sample-specific signal that the model needs to learn from (especially for OOD generalization). Additionally, with 3-way PCGrad active, the gradient projection already manages conflicting gradient directions across in-dist/tandem/extreme-Re groups — accumulating before the step muddies these per-sample group signals.

The halved optimizer step count per epoch (half as many updates for same data) also means the model sees less gradient diversity per epoch, and with cosine_T_max=150 but only 149 epochs completed, the LR schedule wasn't materially different.

### Suggested follow-ups

- **Don't pursue larger accumulation factors** (4x, 8x) — the trend is clearly negative.
- If smoother optimization is desired, **EMA decay tuning** or **Lookahead optimizer** may be better approaches, as they smooth at the parameter level without reducing gradient diversity.
- The result confirms that per-sample gradient noise is actually beneficial for this mesh-based task, consistent with the implicit regularization literature (Smith & Le, 2018).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2335 because ADVISOR: Closing stale PR — gradient accumulation 2x was already reviewed and closed in a prior round. Cleaning up..
